velox/vector/VariantToVector.cpp

```diff
@@ -77,4 +77,165 @@ ArrayVectorPtr variantArrayToVector(
       pool);
 }
 
+// Structure used to keep check on data type of
+// the complex vector, along with the number and data
+// types of constituent elements. Attributes elementCount
+// and elementInserted keeps the length of the vector, and
+// the number of elements inserted during the operation
+// respectively.
+struct NestedTypeCounter {
+  TypeKind kind = TypeKind::UNKNOWN;
+  vector_size_t elementCount = 0;
+  vector_size_t elementsInserted = 0;
+  std::vector<NestedTypeCounter> children;
+};
+
+static void validateOrSetType(NestedTypeCounter& counter, const variant& v) {
+  if (v.isNull()) // If a variant is null, it can be UNKNOWN, INVALID, or
+                  // matching our expected kind
+  {
+    if (v.kind() != TypeKind::UNKNOWN && v.kind() != TypeKind::INVALID &&
```

Comment:
Since we are only checking kind and not the type - this will not work for nested complex types .

Reply:
now checking the entire `Type` tree, thanks for pointing this out!